### PR TITLE
PermissionsAndroid: Request permission only when response from rationale is positive

### PR DESCRIPTION
**Pull request by Bhavik-P**
##  Summary

Currently, the library makes a call to request permission even though the user taps on a neutral or negative response in the rationale dialog after initially denying permission. The mods in this PR will make it so we only make a call to request for permissions when a user taps on a positive response.

## Changelog

[Android][Changed] - Only request permission when the user gives a positive response to the rationale dialog.

## Test Plan
- For the following test cases, Camera permission was used to test.

1. Tap on a button that requests permissions. Approve the permission and make sure the function returns valid permission granted value `true`
-- Reset the permissions after this by either reinstalling or running `adb shell pm reset-permissions`
2. Tap on a button that requests permissions. Deny the permission and make sure the function returns valid permission granted value `false`.
3. Continuation from test case 2, tap on the button again so the rationale dialog pops up with three options. (buttonPositive, buttonNegative, buttonNeutral)
4. Continuation from test case 3, tap on buttonNegative, the dialog disappears and the function returns valid permission granted value `false`.
5. Continuation from test case 4, tap on buttonNeutral, the dialog disappears and the function returns valid permission granted value `false`.
6. Continuation from test case 5, tap on buttonPositive, the dialog disappears and a native android dialog appears to take user's permission.
7. Continuation from test case 6, tap on `Deny`. The native dialog disappears and the function returns a valid permission granted value `false`. This puts you back at 3
8. Tap on `Approve`. The native dialog disappears and the function returns a valid permission granted value `true`.

### CODE REVIEW

**Comment by klazbaba** [comment]
@Bhavik-P, I think this PR is yet to be merged because of the failed tests. Would you mind going through it please?

**Comment by Peeeep** [comment]
The failed tests have nothing to do with this pull request, some maintainer needs to review the request… @satya164 was the original author – could you have a look at this by any chance?

**Comment by ThatMate** [comment]
Will this PR fix the permission dialogue appearing twice when clicked Neutral Button? (explained here: https://github.com/facebook/react-native/issues/28876#issuecomment-729549664 )

**Comment by Peeeep** [comment]
> Will this PR fix the permission dialogue appearing twice when clicked Neutral Button? (explained here: [#28876 (comment)](https://github.com/facebook/react-native/issues/28876#issuecomment-729549664) )

@ThatMate Yes. We use a patch for these changes until it eventually gets merged. Works flawlessly:

```
diff --git a/Libraries/PermissionsAndroid/PermissionsAndroid.js b/Libraries/PermissionsAndroid/PermissionsAndroid.js
index 83124ccc24bd05acac0158f71c73bc4113d84059..35275f097009811a94343abd9f0afee210bc8dda 100644
--- a/Libraries/PermissionsAndroid/PermissionsAndroid.js
+++ b/Libraries/PermissionsAndroid/PermissionsAndroid.js
@@ -216,14 +216,20 @@ class PermissionsAndroid {
           const options = {
             ...rationale,
           };
+          const constants = NativeDialogManagerAndroid.getConstants();
           NativeDialogManagerAndroid.showAlert(
             /* $FlowFixMe(>=0.111.0 site=react_native_fb) This comment
              * suppresses an error found when Flow v0.111 was deployed. To see
              * the error, delete this comment and run Flow. */
             options,
             () => reject(new Error('Error showing rationale')),
-            () =>
-              resolve(NativePermissionsAndroid.requestPermission(permission)),
+            (action, buttonKey) => {
+              if (action === constants.buttonClicked && buttonKey === constants.buttonPositive) {
+                resolve(NativePermissionsAndroid.requestPermission(permission));
+              } else {
+                resolve(this.RESULTS.DENIED);
+              }
+            },
           );
         });
       }
```

**Comment by Bhavik-P** [comment]
> @Bhavik-P, I think this PR is yet to be merged because of the failed tests. Would you mind going through it please?

@klazbaba I don't think the failed tests have anything to do with this PR. If you know someone that can review my code changes and get this through the official pipeline for the upcoming release it would be great. We currently have a patch with this PR and had it in production since I made this PR.

**Comment by sduduzog** [comment]
Would it be possible to expose a way to provide callbacks in instances where the user choice is the neutral button or the negative button. Unless there's already a way to do so, as we already provide custom buttons for those options. i.e. The neutral button has the label "try again later" and as such, we be able to request such a permission after some said time

**Review by HectorRicardo — APPROVED** [review]

**Comment by cjshearer** [comment]
Any update on this? This affects me on v0.61.5, and the changelog makes no mention of later fixes.

**Comment by HectorRicardo** [comment]
FB team: could we get an update about why this pull request hasn't been accepted?

**Comment by SamBaloyi** [comment]
Any update on this?

**Comment by Darkilen** [comment]
Please, can we have an update about this pull request ?

**Comment by Darkilen** [comment]
This PR is still nedeed. Can we please have an update from the team ?

**Pull request closed without merging**
